orttraining/orttraining/core/graph/optimizer_graph_builder.cc

```diff
@@ -443,17 +466,26 @@ Status OptimizerGraphBuilder::BuildInternal(
   // check if all gradients are finite
   ArgDef global_grad_norm_argdef;
   ArgDef global_grad_norm_finite_argdef;
-  if (opt_graph_config_.use_mixed_precision) {
-    //gradient norm for bfloat is not ready yet. skip it to unblock the testing
+
+  if (opt_graph_config_.enable_grad_norm_clip ||
+      (opt_graph_config_.use_mixed_precision &&
+       opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16)) {
+    //gradient norm for bfloat16 is not ready yet. skip it to unblock the testing
     //will add it back when it is ready
-    if (opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16) {
-      ORT_RETURN_IF_ERROR(AddGradientNorm(
-          nodearg_name_generator, gradient_argdefs, graph_defs, global_grad_norm_argdef));
-      optimizer_graph_outputs[OptimizerOutputKey::GlobalGradientNorm] = global_grad_norm_argdef.name;
-      ORT_RETURN_IF_ERROR(AddFiniteGradientCheck(
-          nodearg_name_generator, {global_grad_norm_argdef}, graph_defs, global_grad_norm_finite_argdef));
-      optimizer_graph_outputs[OptimizerOutputKey::GradientAllIsFinite] = global_grad_norm_finite_argdef.name;
-    }
+    ORT_RETURN_IF_ERROR(AddGradientNorm(
+        nodearg_name_generator, gradient_argdefs, graph_defs, global_grad_norm_argdef));
+    optimizer_graph_outputs[OptimizerOutputKey::GlobalGradientNorm] = global_grad_norm_argdef.name;
+  }
+
+  if (opt_graph_config_.use_mixed_precision &&
+      opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16) {
+    ORT_RETURN_IF_ERROR(AddFiniteGradientCheck(
+        nodearg_name_generator, {global_grad_norm_argdef}, graph_defs, global_grad_norm_finite_argdef));
+    optimizer_graph_outputs[OptimizerOutputKey::GradientAllIsFinite] = global_grad_norm_finite_argdef.name;
+  }
+
+  if (global_grad_norm_argdef.name.empty() && global_grad_norm_finite_argdef.name.empty()) {
+    ORT_RETURN_IF_ERROR(AddGradientPassThroughNode(nodearg_name_generator, gradient_argdefs, graph_defs));
   }
```

Comment:
This is tricky. The inputs of `ReduceAllL2` should also be the outputs of `PassThrough`; it means we want to compute the gradient norm from non-updated tensor values. In addition, adding `PassThrough` may increase peak memory. It's better to check the memory consumption using Wei's tool before merging this.

Reply:
WeightUpdate happens after the RedcueAllL2 node, so we won't run into the case you mentioned here. 

So far from what we have seen, peak memory is reached around end of FW. PassThrough is guarding the end of BW, so the memory is probably not a concern here. 
In addition, in other training configs like distributed training and fp16 training, we are already having a barrier node (AllReduce, ReduceAllL2). So introducing the PassThrough here can't make things worse.

Reply:
I agree that connecting PassThrough with ReduceAllL2 doesn't change the math. My intention was making semantic clearer.

I tried WaitEvent for BERT (not sure if it's large or tiny), which is also a kind of Barrier, and adding WaitEvent significantly increases peak memory usage.

---
In reply to: [520924873](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520924873) [](ancestors = 520924873)